redaxo/src/core/lib/setup/setup.php

```diff
@@ -260,6 +260,78 @@ public static function isInitialSetup(): bool
         }
     }
 
+    /**
+     * @return string|false Setup URL or `false` on failure
+     */
+    public static function startWithToken()
+    {
+        $token = rtrim(strtr(base64_encode(random_bytes(32)), '+/', '-_'), '=');
+
+        $configFile = rex_path::coreData('config.yml');
+        $config = rex_file::getConfig($configFile);
+
+        $config['setup'] = isset($config['setup']) && is_array($config['setup']) ? $config['setup'] : [];
+        $config['setup'][$token] = (new DateTimeImmutable('+1 hour'))->format('Y-m-d H:i:s');
+
+        if (false === rex_file::putConfig($configFile, $config)) {
+            return false;
+        }
+
+        return rex_url::backendPage('setup', ['setup_token' => $token], false);
+    }
+
+    public static function isEnabled(): bool
+    {
+        $setup = rex::getProperty('setup', false);
+
+        if (!is_array($setup)) {
+            return (bool) $setup;
+        }
+
+        $currentToken = self::getToken();
+
+        if (!$currentToken && rex::isFrontend()) {
+            return false;
+        }
+
+        $updated = false;
+        foreach ($setup as $token => $expire) {
+            if (strtotime($expire) < time()) {
+                unset($setup[$token]);
+                $updated = true;
+            }
+        }
+
+        if ($updated) {
+            $configFile = rex_path::coreData('config.yml');
+            $config = rex_file::getConfig($configFile);
+            $config['setup'] = $setup ?: false;
+            rex_file::putConfig($configFile, $config);
+        }
+
+        return isset($setup[$currentToken]);
+    }
+
+    public static function getContext(): rex_context
+    {
+        $context = new rex_context([
+            'page' => 'setup',
+            'lang' => rex_request('lang', 'string', ''),
+            'step' => rex_request('step', 'int', 1),
+        ]);
+
+        if ($token = self::getToken()) {
+            $context->setParam('setup_token', $token);
+        }
+
+        return $context;
+    }
+
+    public static function getToken(): ?string
+    {
+        return rex_get('setup_token', 'string', null);
```

Comment:
sollten wir hier das toke validieren?

Reply:
Inzwischen wird die Methode nur noch in der Klasse selbst verwendet, kann sie daher private machen (war in meiner ersten Implementierung anders).
Validieren würde ich darin nicht, da dann unnötig mehrfach validiert werden würde.